core/src/main/scala/chisel3/Printable.scala

```diff
@@ -187,35 +193,47 @@ object FirrtlFormat {
   def unapply(x: Char): Option[Char] =
     Option(x).filter(x => legalSpecifiers contains x)
 
+  // The leading % is optional because this API used to expect just the character
+  // But % is required for newer API when including modifiers and is also supported with just the character
+  private val SpecifierRegex = """(%-?[0-9]*)?([dxbc])""".r
+
   /** Helper for constructing Firrtl Formats
-    * Accepts data to simplify pack
+    *
+    * Accepts data to simplify pack.
+    *
+    * @param specifier the format specifier, e.g. `%0d`
     */
   def apply(specifier: String, data: Data): FirrtlFormat = {
     val bits = data match {
       case b: Bits => b
       case d => throw new Exception(s"Trying to construct FirrtlFormat with non-bits $d!")
     }
     specifier match {
-      case "d" => Decimal(bits)
-      case "x" => Hexadecimal(bits)
-      case "b" => Binary(bits)
-      case "c" => Character(bits)
-      case c   => throw new Exception(s"Illegal format specifier '$c'!")
+      case SpecifierRegex(mod, spec) =>
+        val modifier = if (mod == null) "" else mod.drop(1) // drop leading %
+        spec match {
+          case "d" => Decimal(bits, modifier)
+          case "x" => Hexadecimal(bits, modifier)
+          case "b" => Binary(bits, modifier)
+          case "c" => Character(bits, modifier)
+          case _   => throw new Exception(s"Unreachable! specifier = '$spec'")
+        }
+      case bad => throw new Exception(s"Illegal format specifier '$bad'!")
     }
   }
 }
 
 /** Format bits as Decimal */
-case class Decimal(bits: Bits) extends FirrtlFormat('d')
+case class Decimal(bits: Bits, modifier: String = "") extends FirrtlFormat('d')
```

Comment:
I've realized that this should use richer types, we don't want people writing things like `Decimal(foo, "-4")`, it should be something like `Decimal(foo, Aligment.Left, MinimumDisplayWidth(4))`